src/index.js

```diff
@@ -104,76 +104,97 @@ function rejectOnEmpty(options, result) {
+    , cacheKey = getCacheKey(model, attributes, options);
 
-      return model.findAll(findOptions).then(mapResult.bind(null, attributes, keys, findOptions));
-    }, {
+  if (!GLOBAL_CACHE.has(cacheKey)) {
+    GLOBAL_CACHE.set(cacheKey, loaderForBTM(model, joinTableName, foreignKey, foreignKeyField, {
+      ...options,
       cache: false
     }));
   }
 
-  return cache.get(cacheKey);
+  return GLOBAL_CACHE.get(cacheKey);
 }
 
-function loaderForModel(model, attribute, attributeField, options = {}) {
+function loaderForBTM(model, joinTableName, foreignKey, foreignKeyField, options = {}) {
   assert(options.include === undefined, 'options.include is not supported by model loader');
+  assert(options.association !== undefined, 'options.association should be set for BTM loader');
 
-  let cacheKey = getCacheKey(model, attribute, options);
+  let attributes = [joinTableName, foreignKey];
+  const association = options.association;
+  delete options.association;
 
-  if (!cache.has(cacheKey)) {
-    cache.set(cacheKey, new DataLoader(keys => {
-      const findOptions = Object.assign({}, options);
-      delete findOptions.rejectOnEmpty;
+  return new DataLoader(keys => {
+    let findOptions = Object.assign({}, options);
+    delete findOptions.rejectOnEmpty;
+    if (findOptions.limit) {
+      findOptions.groupedLimit = {
+        through: options.through,
+        on: association,
+        limit: findOptions.limit,
+        values: keys
+      };
+    } else {
+      findOptions.include = [{
+        attributes: [foreignKey],
+        association: association.manyFromSource,
+        where: {
+          [foreignKeyField]: keys,
+          ...options.through.where
+        }
+      }];
+    }
 
-      if (findOptions.limit && keys.length > 1) {
-        findOptions.groupedLimit = {
-          limit: findOptions.limit,
-          on: attributeField,
-          values: keys
-        };
-        delete findOptions.limit;
-      } else {
-        findOptions.where = mergeWhere({
-          [attributeField]: keys
-        }, findOptions.where);
-      }
+    return model.findAll(findOptions).then(mapResult.bind(null, attributes, keys, findOptions));
+  }, {
+    cache: options.cache
+  });
+}
+
+function cachedLoaderForModel(model, attribute, attributeField, options = {}) {
+  assert(options.include === undefined, 'options.include is not supported by model loader');
+
+  let cacheKey = getCacheKey(model, attribute, options);
 
-      return model.findAll(findOptions).then(mapResult.bind(null, attribute, keys, findOptions));
-    }, {
+  if (!GLOBAL_CACHE.has(cacheKey)) {
+    GLOBAL_CACHE.set(cacheKey, loaderForModel(model, attribute, attributeField, {
+      ...options,
       cache: false
     }));
   }
 
-  return cache.get(cacheKey);
+  return GLOBAL_CACHE.get(cacheKey);
+}
+
+function loaderForModel(model, attribute, attributeField, options = {}) {
+  assert(options.include === undefined, 'options.include is not supported by model loader');
```

Comment:
Asserts already checked in cachedLoaderForModel

Reply:
Yes but this method is also called by the shim